# Using Trilateration

In [1]:
import pandas as pd
import numpy as np

##### Define functions

In [2]:
def calculate_distance(RSSI, n, C):
    distance = 10 ** ( (C - RSSI) / (10 * n) )
    return distance

def track(x1,y1,r1,x2,y2,r2,x3,y3,r3):
  A = 2*x2 - 2*x1
  B = 2*y2 - 2*y1
  C = r1**2 - r2**2 - x1**2 + x2**2 - y1**2 + y2**2
  D = 2*x3 - 2*x2
  E = 2*y3 - 2*y2
  F = r2**2 - r3**2 - x2**2 + x3**2 - y2**2 + y3**2
  x = (C*E - F*B) / (E*A - B*D)
  y = (C*D - A*F) / (B*D - A*E)
  return x,y

def calculate_error(x_predict, x_orig, y_predict, y_orig):
    return ( (x_predict - x_orig) ** 2 + (y_predict - y_orig) ** 2 ) ** 0.5

### SCENARIO 1

In [3]:
#read dataset for scenario 1
df_BLE = pd.read_csv('BLE_scenario1.csv')
df_Zigbee = pd.read_csv('Zigbee_scenario1.csv')
df_WiFi = pd.read_csv('WiFi_scenario1.csv')

In [4]:
# Data Preprocessing
BLE_rssi_A= df_BLE['RSSI A']
BLE_rssi_B= df_BLE['RSSI B']
BLE_rssi_C= df_BLE['RSSI C']
BLE_x_true= df_BLE['x']
BLE_y_true= df_BLE['y']

Zigbee_rssi_A= df_Zigbee['RSSI A']
Zigbee_rssi_B= df_Zigbee['RSSI B']
Zigbee_rssi_C= df_Zigbee['RSSI C']
Zigbee_x_true= df_Zigbee['x']
Zigbee_y_true= df_Zigbee['y']

WiFi_rssi_A= df_WiFi['RSSI A']
WiFi_rssi_B= df_WiFi['RSSI B']
WiFi_rssi_C= df_WiFi['RSSI C']
WiFi_x_true= df_WiFi['x']
WiFi_y_true= df_WiFi['y']

x_true= df_BLE['x']
y_true= df_BLE['y']

n_BLE = 2.271
n_Zigbee = 2.935
n_Wifi = 2.162

c_BLE = -75.48
c_Zigbee = -50.33
c_WiFi = -45.73

##### Calculate Distance using Pathloss Model

In [5]:
BLE_distance_A = calculate_distance(BLE_rssi_A,n_BLE,c_BLE)
BLE_distance_B = calculate_distance(BLE_rssi_B,n_BLE,c_BLE)
BLE_distance_C = calculate_distance(BLE_rssi_C,n_BLE,c_BLE)

Zigbee_distance_A = calculate_distance(Zigbee_rssi_A,n_Zigbee,c_Zigbee)
Zigbee_distance_B = calculate_distance(Zigbee_rssi_B,n_Zigbee,c_Zigbee)
Zigbee_distance_C = calculate_distance(Zigbee_rssi_C,n_Zigbee,c_Zigbee)

WiFi_distance_A = calculate_distance(WiFi_rssi_A,n_Wifi,c_WiFi)
WiFi_distance_B = calculate_distance(WiFi_rssi_B,n_Wifi,c_WiFi)
WiFi_distance_C = calculate_distance(WiFi_rssi_C,n_Wifi,c_WiFi)

##### Predict the Coordinates using Trilateration

In [6]:
BLE_x_predict,BLE_y_predict = track(0,0,BLE_distance_A,4,0,BLE_distance_B,0,4,BLE_distance_C)
ZigBee_x_predict,Zigbee_y_predict = track(0,0,Zigbee_distance_A,4,0,Zigbee_distance_B,0,4,Zigbee_distance_C)
WiFi_x_predict,WiFi_y_predict = track(0,0,WiFi_distance_A,4,0,WiFi_distance_B,0,4,WiFi_distance_C)

##### Calculate the error

In [7]:
BLE_error = calculate_error(BLE_x_predict, x_true, BLE_y_predict, y_true)
Zigbee_error = calculate_error(ZigBee_x_predict, x_true, Zigbee_y_predict, y_true)
WiFi_error = calculate_error(WiFi_x_predict, x_true, WiFi_y_predict, y_true)

_error = [] #create error vector
#calcualate error for BLE
BLE_mean_error=np.mean(BLE_error)
_error.append(BLE_mean_error)
#calcualate error for Zigbee
Zigbee_mean_error=np.mean(Zigbee_error)
_error.append(Zigbee_mean_error)
#calcualate error for WiFi
WiFi_mean_error=np.mean(WiFi_error)
_error.append(WiFi_mean_error)

#get average error
average_error = np.mean(_error)
average_error

2.171136408014563

### SCENARIO 2

In [8]:
#read dataset for scenario 2
df2_BLE = pd.read_csv('BLE_scenario2.csv')
df2_Zigbee = pd.read_csv('Zigbee_scenario2.csv')
df2_WiFi = pd.read_csv('WiFi_scenario2.csv')

In [9]:
# Data Preprocessing
BLE2_rssi_A= df2_BLE['RSSI A']
BLE2_rssi_B= df2_BLE['RSSI B']
BLE2_rssi_C= df2_BLE['RSSI C']
BLE2_x_true= df2_BLE['x']
BLE2_y_true= df2_BLE['y']

Zigbee2_rssi_A= df2_Zigbee['RSSI A']
Zigbee2_rssi_B= df2_Zigbee['RSSI B']
Zigbee2_rssi_C= df2_Zigbee['RSSI C']
Zigbee2_x_true= df2_Zigbee['x']
Zigbee2_y_true= df2_Zigbee['y']

WiFi2_rssi_A= df2_WiFi['RSSI A']
WiFi2_rssi_B= df2_WiFi['RSSI B']
WiFi2_rssi_C= df2_WiFi['RSSI C']
WiFi2_x_true= df2_WiFi['x']
WiFi2_y_true= df2_WiFi['y']

x2_true= df2_BLE['x']
y2_true= df2_BLE['y']

n2_BLE = 1.999
n2_Zigbee = 1.912
n2_Wifi = 2.018

c2_BLE = -62.27
c2_Zigbee = -52.73
c2_WiFi = -37.37

##### Calculate Distance using Pathloss Model

In [10]:
BLE2_distance_A = calculate_distance(BLE2_rssi_A,n2_BLE,c2_BLE)
BLE2_distance_B = calculate_distance(BLE2_rssi_B,n2_BLE,c2_BLE)
BLE2_distance_C = calculate_distance(BLE2_rssi_C,n2_BLE,c2_BLE)

Zigbee2_distance_A = calculate_distance(Zigbee2_rssi_A,n2_Zigbee,c2_Zigbee)
Zigbee2_distance_B = calculate_distance(Zigbee2_rssi_B,n2_Zigbee,c2_Zigbee)
Zigbee2_distance_C = calculate_distance(Zigbee2_rssi_C,n2_Zigbee,c2_Zigbee)

WiFi2_distance_A = calculate_distance(WiFi2_rssi_A,n2_Wifi,c2_WiFi)
WiFi2_distance_B = calculate_distance(WiFi2_rssi_B,n2_Wifi,c2_WiFi)
WiFi2_distance_C = calculate_distance(WiFi2_rssi_C,n2_Wifi,c2_WiFi)

##### Predict the Coordinates using Trilateration

In [11]:
BLE2_x_predict,BLE2_y_predict = track(0,0,BLE2_distance_A,3.27,0,BLE2_distance_B,1,4.4,BLE2_distance_C)
ZigBee2_x_predict,Zigbee2_y_predict = track(0,0,Zigbee2_distance_A,3.27,0,Zigbee2_distance_B,1,4.4,Zigbee2_distance_C)
WiFi2_x_predict,WiFi2_y_predict = track(0,0,WiFi2_distance_A,3.27,0,WiFi2_distance_B,1,4.4,WiFi2_distance_C)

##### Calculate the error

In [12]:
BLE2_error = calculate_error(BLE2_x_predict, x2_true, BLE2_y_predict, y2_true)
Zigbee2_error = calculate_error(ZigBee2_x_predict, x2_true, Zigbee2_y_predict, y2_true)
WiFi2_error = calculate_error(WiFi2_x_predict, x2_true, WiFi2_y_predict, y2_true)

_error2 = [] #create error vector
#calcualate error for BLE
BLE2_mean_error=np.mean(BLE2_error)
_error2.append(BLE2_mean_error)
#calcualate error for Zigbee
Zigbee2_mean_error=np.mean(Zigbee2_error)
_error2.append(Zigbee2_mean_error)
#calcualate error for WiFi
WiFi2_mean_error=np.mean(WiFi2_error)
_error2.append(WiFi2_mean_error)

#get average error
average_error2 = np.mean(_error2)
average_error2

3.281451266978312

### SCENARIO 3

In [13]:
#read dataset for scenario 2
df3_BLE = pd.read_csv('BLE_scenario3.csv')
df3_Zigbee = pd.read_csv('Zigbee_scenario3.csv')
df3_WiFi = pd.read_csv('WiFi_scenario3.csv')

In [14]:
# Data Preprocessing
BLE3_rssi_A= df3_BLE['RSSI A']
BLE3_rssi_B= df3_BLE['RSSI B']
BLE3_rssi_C= df3_BLE['RSSI C']
BLE3_x_true= df3_BLE['x']
BLE3_y_true= df3_BLE['y']

Zigbee3_rssi_A= df3_Zigbee['RSSI A']
Zigbee3_rssi_B= df3_Zigbee['RSSI B']
Zigbee3_rssi_C= df3_Zigbee['RSSI C']
Zigbee3_x_true= df3_Zigbee['x']
Zigbee3_y_true= df3_Zigbee['y']

WiFi3_rssi_A= df3_WiFi['RSSI A']
WiFi3_rssi_B= df3_WiFi['RSSI B']
WiFi3_rssi_C= df3_WiFi['RSSI C']
WiFi3_x_true= df3_WiFi['x']
WiFi3_y_true= df3_WiFi['y']

x3_true= df3_BLE['x']
y3_true= df3_BLE['y']

n3_BLE = 2.442
n3_Zigbee = 2.085
n3_Wifi = 2.563

c3_BLE = -62.5
c3_Zigbee = -48.52
c3_WiFi = -33.75

##### Calculate Distance using Pathloss Model

In [15]:
BLE3_distance_A = calculate_distance(BLE3_rssi_A,n3_BLE,c3_BLE)
BLE3_distance_B = calculate_distance(BLE3_rssi_B,n3_BLE,c3_BLE)
BLE3_distance_C = calculate_distance(BLE3_rssi_C,n3_BLE,c3_BLE)

Zigbee3_distance_A = calculate_distance(Zigbee3_rssi_A,n3_Zigbee,c3_Zigbee)
Zigbee3_distance_B = calculate_distance(Zigbee3_rssi_B,n3_Zigbee,c3_Zigbee)
Zigbee3_distance_C = calculate_distance(Zigbee3_rssi_C,n3_Zigbee,c3_Zigbee)

WiFi3_distance_A = calculate_distance(WiFi3_rssi_A,n3_Wifi,c3_WiFi)
WiFi3_distance_B = calculate_distance(WiFi3_rssi_B,n3_Wifi,c3_WiFi)
WiFi3_distance_C = calculate_distance(WiFi3_rssi_C,n3_Wifi,c3_WiFi)

##### Predict the Coordinates using Trilateration

In [16]:
BLE3_x_predict,BLE3_y_predict = track(0,0,BLE3_distance_A,9.625,0,BLE3_distance_B,4.812,2.492,BLE3_distance_C)
ZigBee3_x_predict,Zigbee3_y_predict = track(0,0,Zigbee3_distance_A,9.625,0,Zigbee3_distance_B,4.812,2.492,Zigbee3_distance_C)
WiFi3_x_predict,WiFi3_y_predict = track(0,0,WiFi3_distance_A,9.625,0,WiFi3_distance_B,4.812,2.492,WiFi3_distance_C)

##### Calculate the error

In [17]:
BLE3_error = calculate_error(BLE3_x_predict, x3_true, BLE3_y_predict, y3_true)
Zigbee3_error = calculate_error(ZigBee3_x_predict, x3_true, Zigbee3_y_predict, y3_true)
WiFi3_error = calculate_error(WiFi3_x_predict, x3_true, WiFi3_y_predict, y3_true)

_error3 = [] #create error vector
#calcualate error for BLE
BLE3_mean_error=np.mean(BLE3_error)
_error3.append(BLE3_mean_error)
#calcualate error for Zigbee
Zigbee3_mean_error=np.mean(Zigbee3_error)
_error3.append(Zigbee3_mean_error)
#calcualate error for WiFi
WiFi3_mean_error=np.mean(WiFi3_error)
_error3.append(WiFi3_mean_error)

#get average error
average_error3 = np.mean(_error3)
average_error3

5.4920977771990716

## Summary

<img src="Trilateration.jpg"/>

According to the results, Scenario 1 got the least  average error with 2.17 meters. Scenario 2 got an average error of 3.28 meters for all technologies. Scenario 3 got the highest average error of 5.49 meters for all wireless technologies. 

Overall, the trilateration method deviated off from the original receiver position of 3.648228422 meters